events/handler.go

```diff
@@ -0,0 +1,180 @@
+	LastNEvents int
+}
+
+// EventCategory is an enumerated type which lists the categories under which
+// events may fall. The Event field EventType is populated by this enum.
+type EventCategory int
+
+const (
+	OomInstanceCat EventCategory = iota
+	ContainerCreation
+	ContainerDeletion
+)
+
+// a general interface which populates the Event field EventData. The actual
+// object, such as an OomInstance, is set as an Event's EventData
+type EventDataInterface interface {
+}
+
+// returns a pointer to an initialized Events object
+func NewEventManager() *events {
+	return &events{
+		eventlist: make(EventSlice, 0),
+	}
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+	return &Request{
+		EventType: map[EventCategory]bool{},
+	}
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Len() int {
+	return len(e)
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Less(i, j int) bool {
+	return e[i].TimeOfEvent.Before(e[j].TimeOfEvent)
+}
+
+// sorts and returns up to the last N chronological elements
+func getLastNEvents(request *Request, eSlice EventSlice) EventSlice {
+	sort.Sort(eSlice)
+	n := request.LastNEvents
+	if n >= eSlice.Len() || n <= 0 {
+		return eSlice
+	}
+	return eSlice[eSlice.Len()-n:]
+}
+
+// determines if an event occurs within the time set in the request object and is the right type
+func checkIfEventSatisfiesRequest(request *Request, event *Event) bool {
+	startTime := request.StartTime
+	endTime := request.EndTime
+	eventTime := event.TimeOfEvent
+	if !startTime.IsZero() {
+		if startTime.After(eventTime) {
+			glog.Infof("start time ins't o")
+			return false
+		}
+	}
+	if !endTime.IsZero() {
+		if endTime.Before(eventTime) {
+			glog.Infof("end time isn't zero")
+			return false
+		}
+	}
+	if request.EventType[event.EventType] != true {
+		glog.Infof("event type is not found")
+		return false
+	}
+	return true
+}
+
+// method of Events object that screens Event objects found in the eventlist
+// attribute and if they fit the parameters passed by the Request object,
+// adds it to a slice of *Event objects that is returned
+func (self *events) GetEvents(request *Request) (EventSlice, error) {
+	returnEventList := EventSlice{}
+	for _, e := range self.eventlist {
+		if checkIfEventSatisfiesRequest(request, e) {
+			returnEventList = append(returnEventList, e)
+		}
+	}
+	returnEventList = getLastNEvents(request, returnEventList)
+	return returnEventList, nil
+}
+
+// method of Events object that maintains an *Event channel passed by the user.
+// When an event occurs that satisfies the parameters in the passed Request
+// object it is fed to the channel
+func (self *events) WatchEvents(eventChannel chan *Event) error {
```

Comment:
There are a few different ways WatchEvents can be implemented depending on how it’s used

1. func (self *events) WatchEvents(eventChannel chan *Event) error
The current implementation.  WatchEvents is called in a goroutine in the manager when the manager is started.  The EventManager is a field of the manager struct.  When events occur, they are fed to the eventChannel.  WatchEvents adds the events it receives in the channel to the EventManager.  AddEvents and WatchEvents are redundant here.

2. func (self *events) WatchEvents(eventChannel chan *Event, request *Request) error
WatchEvents is called in a goroutine in the manager when the manager is started.  The EventManager is a field of the manager struct.  When events occur, they are fed to the eventChannel.  In this implementation, WatchEvents would only add events to the eventlist of self if they satisfy the request. To make this effective, one of two things would have to occur.
a. The caller has multiple EventManager objects, each of which differs by request object.  The manager struct would have a field map[*Request]*events. It creates a unique channel and spins up a different goroutine for each WatchEvents call with a different Request argument for each
b. The *events object that implements EventManager has a collection of eventlists rather than one. The field would look like map[*Request]EventSlice.

This has the potential downfall of storing every event multiple times in every eventlist corresponding to all the Requests the caller asks for that it satisfies.

3. func (self *events) WatchEvents(inChannel chan *Event, outChannel chan *Event, request *Request) error
The caller creates an input and output channel. WatchEvents is called in a goroutine in the manager when the manager is started.  The EventManager is a field of the manager struct.  When events occur, they are fed to the inChannel.  If they satisfy the request, they are fed back to the outChannel and received by the caller.  Then the events may be added by the caller to the EventManager held by the manager struct.

4. WatchEvents is not implemented in the events handler. A goroutine is called when manager starts that has a channel it reads from.  When events are created after either watchForNewContainers or watchForNewOoms, both in manager, they are fed to the channel. Within the goroutine AddEvents is called to update the manager struct’s EventManager.

5. The oomparser and container libraries implement the 3 functions of the EventManager interface and *events and its implementations are removed from the events handler.

4 seems like the least redundant solution. Because space is not an issue, 2.b. seems like it could be a good choice.  3 is the closest to the previous commit that was on github, with the main change that events no longer depends on oomparser.